### PR TITLE
[linux-port] Casts for compares and assigns

### PR DESCRIPTION
**Pull request by pow2clk**
Various comparisons produce warnings because the values differ in
signedness. This either changes the type of one of the parameters
or else it casts one. This includes adding casts to all of the
defined return error values. Which is more than was needed, but
consistent with some existing values anyway.
Fixes 11 clang and 12 gcc warnings.

Eliminate a few greater than or equal comparisons of unsigned
values which are always true. The most aggressive of this is
the exclusion of a function that has become a no-op because of
HLSL initialization of a key variable that had some invalid
comparisons because of the same const value.
Fixes 4 clang and 7 gcc warnings.

Additionally casts a few parameters to the proper type to silence
warnings about confusing or problem casts. This includes printing
functions, array indices, and deletes of void pointers. The latter
of which might have caused problems.
Fixes 6 clang 6 gcc warnings.

One incidental change that silences a warning about security for
using a string variable as the format for printf. Since it just
outputs the string, fputs is more appropriate anyway. Also
directed to stderr as more appropriate for debug info.

Contributes to https://github.com/google/DirectXShaderCompiler/issues/206

### CODE REVIEW

File: lib/HLSL/DxilRootSignature.cpp [hunk 1]
```diff
@@ -207,7 +207,7 @@ HRESULT SimpleSerializer::ReserveBlock(void **ppData, unsigned cbSize,
 
 Cleanup:
   if (FAILED(hr)) {
-    delete[] pClonedData;
+    delete[] (char*)pClonedData;
```
Comment (ehsannas) [inline lib/HLSL/DxilRootSignature.cpp]: according to [this](http://www.cplusplus.com/reference/new/operator%20delete[]/), `operator delete []` does take a `void*` 
pClonedData is also a void* as far as I can tell. I'm wondering why this cast is needed?

File: lib/HLSL/DxilRootSignature.cpp [hunk 2]
```diff
@@ -146,7 +146,7 @@ SimpleSerializer::~SimpleSerializer() {
     m_pSegment = pSegment->pNext;
 
     if (pSegment->bOwner) {
-      delete pSegment->pData;
+      delete[] (char*)pSegment->pData;
```
Comment (ehsannas) [inline lib/HLSL/DxilRootSignature.cpp]: why is this necessary? `delete` on a `void*` is valid, is it not?

File: lib/HLSL/DxilRootSignature.cpp (hunk 1 above)
Comment (pow2clk) [reply lib/HLSL/DxilRootSignature.cpp]: The linked reference does indicate a void pointer, but I think that's to indicate that any pointer is valid. They don't really mean that a void pointer is a good idea in all cases. The canonical example is when a class has a destructor. If you delete it as a void, the destructor doesn't get called. [This](https://www.viva64.com/en/w/v772/) gives a pretty succinct summary and even cites the spec, which seems like a great way to add credibility to an argument. Especially when most people haven't paid up to get access to it. You could say it says anything! 

That said, we are using PODs here. So maybe the compiler is protesting too much. I don't see much point in the pointer ever being void. I'll try a few things and better nail down the specific warning and alternatives.

File: lib/HLSL/DxilRootSignature.cpp (hunk 2 above)
Comment (pow2clk) [reply lib/HLSL/DxilRootSignature.cpp]: In addition to the caveats above, in this case, the allocation was done as an array. There's a good chance the compiler does the right thing with PODs, but using `delete` instead of `delete[]` is clearly a violation of the letter of the law.

Comment (ehsannas) [reply lib/HLSL/DxilRootSignature.cpp]: Yes. Agreed.

**Comment by pow2clk** [comment]
Regarding void pointer deletes, here is the warning clang throws:
>warning: cannot delete expression with pointer-to-'void' type 'void *' [-Wdelete-incomplet]
    delete[] pClonedData;
GCC protests as well:
> warning: deleting 'void*' is undefined [-Wdelete-incomplete]

While there's a good chance that the compilers give out exasperated sighs and then do the right thing, I don't think there's much harm in humoring them. Though it might look cleaner if I just got rid of the void pointer entirely.

**Comment by ehsannas** [comment]
The explanations made things clearer for me. I think the current solution is clean enough to be merged.
Thanks

**Comment by pow2clk** [comment]
Well. I'm reluctant to change the pointers to char because they are used to pass around classes like `DxilContainerRootParameter`. Maybe in the spirit of reflection, since these were cast allocated as arrays of char, they should be deleted the same way?

**Comment by pow2clk** [comment]
Oh! great. the merge passed me in the night :smile:

**Comment by ehsannas** [comment]
Ah! sorry. prematurely merged. Will roll back.

**Comment by ehsannas** [comment]
https://github.com/google/DirectXShaderCompiler/pull/230 for rolling this back.
I'm afraid you need to open a new PR. Sorry about that.
